### PR TITLE
BUG: require arraylike in infer_dtype_from_array

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Comment by jreback** [comment]
hmm 32-bit error legit?

**Comment by jbrockmendel** [comment]
> hmm 32-bit error legit?

I think so, still troubleshooting it (and similar in #38505).  in the meantime ill shamelessly direct your attention to #38504, #34683, #37965, #37950, #38021, #38073, #38285, #36131, #35417 (the latter two being harder than the rest)

**Comment by jreback** [comment]
> > hmm 32-bit error legit?
> 
> I think so, still troubleshooting it (and similar in #38505). in the meantime ill shamelessly direct your attention to #38504, #34683, #37965, #37950, #38021, #38073, #38285, #36131, #35417 (the latter two being harder than the rest)

looking now

File: pandas/tests/dtypes/cast/test_infer_dtype.py [hunk 1]
```diff
@@ -141,12 +141,29 @@ def test_infer_dtype_from_scalar_errors():
...
     [
         ("foo", np.object_, False),
         (b"foo", np.object_, False),
-        (1, np.int_, False),
+        (1, np.int64, False),
         (1.5, np.float_, False),
+        (np.datetime64("2016-01-01"), np.dtype("M8[ns]"), False),
+        (Timestamp("20160101"), np.dtype("M8[ns]"), False),
+        (Timestamp("20160101", tz="UTC"), np.object_, False),
+        (Timestamp("20160101", tz="UTC"), "datetime64[ns, UTC]", True),
```
Comment (jreback) [inline pandas/tests/dtypes/cast/test_infer_dtype.py:153]: should we be return a DatetimeTZDtype object?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/dtypes/cast/test_infer_dtype.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/tests/dtypes/cast/test_infer_dtype.py:153]: we do

Comment (jreback) [reply pandas/tests/dtypes/cast/test_infer_dtype.py:153]: > "datetime64[ns, UTC]" is a string, aren't we *specifically* return a DatetimeTZDtype object?

Comment (jbrockmendel) [reply pandas/tests/dtypes/cast/test_infer_dtype.py:153]: yes, but this test uses is_dtype_equal to compare the result with expected

Comment (jreback) [reply pandas/tests/dtypes/cast/test_infer_dtype.py:153]: ok (we did merge a similar patch for categorical so ok)

**Review by jreback — APPROVED** [review]

**Pull request merged**